### ISSUE TITLE
Illegal SharedPreferences filename on Windows

### ISSUE THREAD

**Issue by drizzd** [robolectric/robolectric#5529]
### Description

I am trying to unit test [Firestore](https://firebase.google.com/docs/firestore/). Unfortunately, they have a `:` in their SharedPreferences filename: https://github.com/firebase/firebase-android-sdk/blob/master/firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java#L46. Since this character is not allowed on Windows, initializing the app fails with the following error:

```
java.lang.IllegalStateException: java.lang.RuntimeException: java.io.IOException: The filename, directory name, or volume label syntax is incorrect

	at android.app.SharedPreferencesImpl.awaitLoadedLocked(SharedPreferencesImpl.java:265)
	at android.app.SharedPreferencesImpl.edit(SharedPreferencesImpl.java:349)
	at com.example.RobolectricTests.sharedPreferencesTest(MyTest.kt:15)
[...]
```

### Steps to Reproduce

Running this unit test demonstrates the issue:

```kt
package com.example

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class RobolectricTests {
    @Test
    fun sharedPreferencesTest() {
        val context = ApplicationProvider.getApplicationContext<Context>()
        val sharedPreferences = context.getSharedPreferences("com.example:foo", 0)
        sharedPreferences.edit().putString("testId", "12345").commit()
    }
}
```

### Robolectric & Android Version

```
android {
    compileSdkVersion 28
[...]
}

dependencies {
[...]
    testImplementation 'junit:junit:4.12'
    testImplementation 'androidx.test:core:1.2.0'
    testImplementation 'androidx.test.ext:junit:1.1.1'
    testImplementation 'androidx.test:runner:1.2.0'
    testImplementation 'org.robolectric:robolectric:4.3.1'
    testImplementation 'org.jetbrains.kotlinx:kotlinx-coroutines-test:1.3.2'
[...]
```

### Link to a public git repo demonstrating the problem:

I would have to set this up. I hope the above will be sufficient to reproduce the issue.

**Comment by drizzd** [comment]
The issue goes away if I downgrade to `robolectric:4.2`, same version they use to test the [Firebase SDK](https://github.com/firebase/firebase-android-sdk/blob/be3d80627383ba5ad9792db991de13108430715b/firebase-firestore/firebase-firestore.gradle#L143-L145) currently.

**Comment by drizzd** [comment]
The root cause for the difference between 4.2 and 4.3.1 is 2d10c50a07a02ae129f94d7d87201ed5e923025a. In android.app.SharedPreferencesImpl we call Os.stat. Os.stat returns `StructStat.st_mode=0` in 4.2. In 4.3 we call `OsConstantsValues.getMode(path)` which accesses the file system and crashes.

```
java.io.IOException: The filename, directory name, or volume label syntax is incorrect
	at java.io.WinNTFileSystem.canonicalize0(Native Method)
	at java.io.WinNTFileSystem.canonicalize(WinNTFileSystem.java:428)
	at java.io.File.getCanonicalPath(File.java:618)
	at org.robolectric.shadows.OsConstantsValues.canonicalize(OsConstantsValues.java:58)
	at org.robolectric.shadows.OsConstantsValues.getMode(OsConstantsValues.java:50)
	at org.robolectric.shadows.ShadowLinux.stat(ShadowLinux.java:30)
	at libcore.io.Linux.stat(Linux.java)
	at libcore.io.BlockGuardOs.$$robo$$libcore_io_BlockGuardOs$stat(BlockGuardOs.java:328)
	at libcore.io.BlockGuardOs.stat(BlockGuardOs.java)
	at android.system.Os.$$robo$$android_system_Os$stat(Os.java:565)
	at android.system.Os.stat(Os.java)
	at android.app.SharedPreferencesImpl.$$robo$$android_app_SharedPreferencesImpl$loadFromDisk(SharedPreferencesImpl.java:147)
	at android.app.SharedPreferencesImpl.loadFromDisk(SharedPreferencesImpl.java)
	at android.app.SharedPreferencesImpl.access$000(SharedPreferencesImpl.java:54)
	at android.app.SharedPreferencesImpl$1.$$robo$$android_app_SharedPreferencesImpl_1$run(SharedPreferencesImpl.java:122)
	at android.app.SharedPreferencesImpl$1.run(SharedPreferencesImpl.java)
```

**Comment by drizzd** [comment]
Duplicate of #5454.

**Comment by monashreer** [comment]
Can someone look into this issue? If it can be fixed in upcoming version?

**Comment by hoisie** [comment]
@monashreer this has been fixed, see #5454. It is available in 4.5-alpha-1.
